torch/jit/_recursive.py

```diff
@@ -238,3 +518,56 @@ def create_constant_iterable_module(module):
     else:
         raise RuntimeError("Only nn.ModuleList, nn.Sequential, and nn.ModuleDict can be made "
                            "into constant modules, found {}".format(module))
+
+def wrap_cpp_module(cpp_module):
+    """
+    Wrap this torch._C.ScriptModule in a Python ScriptModule, recursively for all submodules
+    """
+    script_module = torch.jit.RecursiveScriptModule(cpp_module)
+    for name, cpp_module in script_module._c._get_modules():
+        setattr(script_module, name, wrap_cpp_module(cpp_module))
+
+    script_module._finalize()
+    return script_module
+
+def compile_unbound_method(concrete_type, fn):
+    if _jit_internal.is_ignored_fn(fn):
+        return None
+    stub = make_stub(fn)
+    with torch.jit._disable_emit_hooks():
+        # We don't want to call the hooks here since the graph that is calling
+        # this function is not yet complete
+        create_methods_from_stubs(concrete_type, (stub,))
+    return stub
+
+def lazy_bind(concrete_type, unbound_method):
+    """
+    Returns a function that lazily binds `unbound_method` to a provided
+    Module IValue, then invokes the method. We do this so that any Python
+    shenanigans that will poison type sharing are impossible at compiled
+    time.
+    """
+    def lazy_binding_method(cpp_module, *args):
+        script_module = torch.jit.RecursiveScriptModule(cpp_module)
+        orig_class = concrete_type.py_class
+
+        # Copy @ignored/@unused methods from the original module to the new one.
+        # This ensures they are available during compilation.
+        for name in dir(orig_class):
+            item = getattr(orig_class, name, None)
+            if _jit_internal.is_ignored_fn(item):
+                setattr(script_module, name, item)
+
+        for name, value in concrete_type.get_constants().items():
+            setattr(script_module, name, value)
+
+        script_module._finalize()
```

Comment:
Let's refactor so that there is one pathway that goes from ivalue + python class -> python object for modules. This is used here, when originally constructing the pyobject, and on torch.jit.load.